pkg/autodiscovery/providers/prometheus_common.go

```diff
@@ -0,0 +1,226 @@
+	TLSHostHeader                 bool                        `mapstructure:"tls_use_host_header" json:"tls_use_host_header,omitempty"`
+	TLSIgnoreWarn                 bool                        `mapstructure:"tls_ignore_warning" json:"tls_ignore_warning,omitempty"`
+	TLSCert                       string                      `mapstructure:"tls_cert" json:"tls_cert,omitempty"`
+	TLSPrivateKey                 string                      `mapstructure:"tls_private_key" json:"tls_private_key,omitempty"`
+	TLSCACert                     string                      `mapstructure:"tls_ca_cert" json:"tls_ca_cert,omitempty"`
+	Headers                       map[string]string           `mapstructure:"headers" json:"headers,omitempty"`
+	ExtraHeaders                  map[string]string           `mapstructure:"extra_headers" json:"extra_headers,omitempty"`
+	Timeout                       int                         `mapstructure:"timeout" json:"timeout,omitempty"`
+	Tags                          []string                    `mapstructure:"tags" json:"tags,omitempty"`
+	Service                       string                      `mapstructure:"service" json:"service,omitempty"`
+	MinCollectInterval            int                         `mapstructure:"min_collection_interval" json:"min_collection_interval,omitempty"`
+	EmptyDefaultHost              bool                        `mapstructure:"empty_default_hostname" json:"empty_default_hostname,omitempty"`
+}
+
+// LabelJoinsConfig contains the label join configuration fields
+type LabelJoinsConfig struct {
+	LabelsToMatch []string `mapstructure:"labels_to_match" json:"labels_to_match"`
+	LabelsToGet   []string `mapstructure:"labels_to_get" json:"labels_to_get"`
+}
+
+// ADConfig contains the autodiscovery configuration data for a PrometheusCheck
+type ADConfig struct {
+	ExcludeAutoconf    *bool     `mapstructure:"exclude_autoconfig_files"`
+	KubeAnnotations    *InclExcl `mapstructure:"kubernetes_annotations"`
+	KubeContainerNames []string  `mapstructure:"kubernetes_container_names"`
+	containersRe       *regexp.Regexp
+}
+
+// InclExcl contains the include/exclude data structure
+type InclExcl struct {
+	Incl map[string]string `mapstructure:"include"`
+	Excl map[string]string `mapstructure:"exclude"`
+}
+
+// defaultCheck has the default openmetrics check values
+// To be used when the checks configuration is empty
+var defaultCheck = &PrometheusCheck{
+	Instances: []*OpenmetricsInstance{
+		{
+			Metrics:   openmetricsDefaultMetrics,
+			Namespace: openmetricsDefaultNS,
+		},
+	},
+	AD: &ADConfig{
+		ExcludeAutoconf: boolPointer(true),
+		KubeAnnotations: &InclExcl{
+			Excl: map[string]string{prometheusScrapeAnnotation: "false"},
+			Incl: map[string]string{prometheusScrapeAnnotation: "true"},
+		},
+		KubeContainerNames: []string{},
+	},
+}
+
+// init prepares the PrometheusCheck structure and defaults its values
+// init must be called only once
+func (pc *PrometheusCheck) init() error {
```

Comment:
Hi @truthbk, to be fair @ahmed-mez did have the `sync.Once` used in earlier commits here :) but since the initialization of these is controlled by `PrometheusPodsConfigProvider. setupConfigs` we agreed that it made sense to get rid of it. However maybe a better way would be to move this initialization to be implemented as helpers in `PrometheusPodsConfigProvider`. The need to have an `init()` arises from the fact that there isn't a way to perform this initialization when `config.Datadog.UnmarshalKey` is called.

You can see the original discussion around this [here](https://github.com/DataDog/datadog-agent/pull/6358#discussion_r488103219).

Reply:
make sens.